### PR TITLE
[MRG] ENH: Add function to search for specific connections

### PR DESCRIPTION
**Pull request by ntolley**
This PR turned into a gauntlet of if/else statements so I'd definitely love some input if you guys see a place to simplify. In any case I am quite happy with the functionality and to my eyes everything seems to be working correctly. The API works as follows:
```py
net.pick_connection(src_gids='L2_pyramidal', target_gids=[0, 1, 2])
```
Will return the index of any connections in `net.connectivity` that _both_ contain L2 pyramidal cells as `src_gids`, and `[0, 1, 2]` in the set of `target_gids`. 

```py
net.pick_connection(receptor=['nmda', 'ampa'])
```
Will return all connections that have _either_ NMDA or AMPA synapses.

More succinctly, supplying a list to a single parameter corresponds to a `logical_or`, whereas supplying multiple parameters corresponds to a `logical_and` across the parameters. While it's wordy to explain, I find this setup to be the most intuitive with what I expect in a search function.

Let me know what you guys think!

### CODE REVIEW

**Comment by jasmainak** [comment]
> More succinctly, supplying a list to a single parameter corresponds to a logical_or, whereas supplying multiple parameters corresponds to a logical_and across the parameters. While it's wordy to explain, I find this setup to be the most intuitive with what I expect in a search function.

I think this makes sense to me!

File: hnn_core/network.py [hunk 1]
```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
...
+            valid loc='soma'.
+        receptor : str | list of str | None
+            Synaptic receptor of connection. Must be one of:
+            'ampa', 'nmda', 'gabaa', or 'gabab'.
+
+        Returns
+        -------
+        conn_indices : list of int
+            List of indices corresponding to items in net.connectivity.
+            Connecion indices are added if any of the provided parameter
```
Comment (jasmainak) [inline hnn_core/network.py:1219]:
```suggestion
            Connecion indices are included if any of the provided parameter
```

somehow on a quick read, "added" gave me the impression that `net.connectivity` had a new item

File: hnn_core/network.py [hunk 2]
```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
...
+            inner_set = set()
+            # Union of indices which match inputs for single parameter
+            for term in search_terms:
+                inner_set = inner_set.union(search_dict[term])
+            # Intersection across parameters
+            if conn_set and inner_set:
+                conn_set = conn_set.intersection(inner_set)
+                if not conn_set:  # Exit if both are non empty disjoint sets
+                    raise Warning('No connections match search parameters.'
+                                  'Try changing or excluding parameters.')
```
Comment (jasmainak) [inline hnn_core/network.py:1335]: I would just return an empty list without complaining. Warnings can be very notorious in terms of polluting the output log, specially if the user is running something in a loop and this condition happens often

File: hnn_core/tests/test_network.py [hunk 3]
```diff
@@ -250,6 +250,65 @@ def test_network():
...
+        ('target_gids', 35, [22, 34, 35, 37, 38, 40, 41, 42, 43, 44]),
+        ('target_gids', range(2), [30, 39]),
+        ('target_gids', 'L2_pyramidal',
+         [22, 34, 35, 37, 38, 40, 41, 42, 43, 44]),
+        ('target_gids', None,
+         [22, 28, 30, 34, 35, 37, 38, 39, 40, 41, 42, 43, 44]),
+        ('loc', 'soma', [30, 39]),
+        ('loc', None, [30, 39]),
+        ('receptor', 'gabaa', [30, 39]),
+        ('receptor', None, [30, 39])]
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:272]: i'm a bit afraid that this hardcoded test will break if something changes the order of the `connectivity` list. Another option is check if all the returned indices indeed contain the match terms and the indices that are not returned don't contain the match term.

File: examples/plot_connectivity.py [hunk 4]
```diff
@@ -124,13 +127,14 @@
...
-src_gid = net_sparse.connectivity[conn_idx]['src_range'][gid_idx]
+gid_idx = 5
+conn_idx = net_sparse.pick_connection(src_gids='L2_basket')[-1]
+src_gid = net_sparse.connectivity[conn_idx]['src_gids'][gid_idx]
 plot_connectivity_matrix(net_sparse, conn_idx)
 plot_cell_connectivity(net_sparse, conn_idx, src_gid)
 
-conn_idx, gid_idx = len(net_sparse.connectivity) - 2, 5
-src_gid = net_sparse.connectivity[conn_idx]['src_range'][gid_idx]
+conn_idx -= 1
```
Comment (jasmainak) [inline examples/plot_connectivity.py:136]: this still feels a bit contrived ...

**Comment by jasmainak** [comment]
I like this @ntolley !!!

File: hnn_core/tests/test_network.py (hunk 3 above)
Comment (ntolley) [reply hnn_core/tests/test_network.py:272]: I agree for most situations, but for this part of the codebase I would really want to know if something changes the order of `net.connectivity`. Fragile tests like 
https://github.com/jonescompneurolab/hnn-core/blob/30fb2a15d02760a7785404b88182683be5b4e455/hnn_core/tests/test_network.py#L256
were really useful for catching that bug in the drives API.

If you forsee a lot of situations where this might crop up and be a pain then I'm totally down to change.

File: hnn_core/network.py [hunk 5]
```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
...
+            src_gids = [src_gids]
+        elif isinstance(src_gids, str):
+            _check_option('src_gids', src_gids, valid_srcs)
+            src_gids = self.gid_ranges[_long_name(src_gids)]
+        for src_gid in src_gids:
+            _validate_type(src_gid, int, 'src_gid')
+            gid_type = self.gid_to_type(src_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'src_gid {src_gid} not in net.gid_ranges')
```
Comment (jasmainak) [inline hnn_core/network.py:1261]: can some of this be refactored with `add_connection`? I think it's time we created a file called `check.py` which has all these checks. See MNE and scikit-learn for inspiration:

https://github.com/mne-tools/mne-python/blob/main/mne/utils/check.py
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/validation.py

File: hnn_core/network.py [hunk 6]
```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
...
+            target_gids = [target_gids]
+        elif isinstance(target_gids, str):
+            _check_option('target_gids', target_gids, valid_targets)
+            target_gids = self.gid_ranges[_long_name(target_gids)]
+        for target_gid in target_gids:
+            _validate_type(target_gid, int, 'target_gid')
+            gid_type = self.gid_to_type(target_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'target_gid {src_gid} not in net.gid_ranges')
```
Comment (jasmainak) [inline hnn_core/network.py:1276]: do:

```py
target_gids = _check_gids(target_gids)
```

or some such thing

**Comment by jasmainak** [comment]
@ntolley I recommend changing the title back to [WIP] when you are not actively seeking reviews or waiting for the PR to be merged. These prefixes allow us to filter the PRs that need attention: https://github.com/jonescompneurolab/hnn-core/pulls?q=is%3Apr+is%3Aopen+MRG. Probably not a biggie here but more important in larger projects that you may want to contribute to in the future :)

File: hnn_core/network.py (hunk 5 above)
Comment (ntolley) [reply hnn_core/network.py]: Created a `check,py` and moved a lot of tests there. In `add_connection` the `target_gids` still have some dedicated checks since they are turned into a more complicated structure, but in general it reduced a lot of redundancy.

**Comment by ntolley** [comment]
Ready for more reviews here. The major change is the addition of `check.py` which cleans up a lot of redundant code in both `net.add_connection`, and `net.pick_connection`. I'll have to do a sweep to see if there's anything else that can benefit from the existing functions, or can be simplified by moving generic input checking to this file.

File: hnn_core/network.py [hunk 7]
```diff
@@ -20,6 +20,7 @@
 from .viz import plot_cells
 from .externals.mne import _validate_type, _check_option
 from .extracellular import ExtracellularArray
+from .check import (_create_gid_list, _gid_to_type, _string_input_to_list)
```
Comment (jasmainak) [inline hnn_core/network.py:23]:
```suggestion
from .check import _create_gid_list, _gid_to_type, _string_input_to_list
```

File: hnn_core/check.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+"""Input check functions."""
+
+# Authors: Nick Tolley <nicholas_tolley@brown.edu>
+
+from .params import _long_name
+from .externals.mne import _validate_type, _check_option
+
+
+def _create_gid_list(gids, gid_ranges, valid_cells, arg_name):
```
Comment (jasmainak) [inline hnn_core/check.py:9]:
```suggestion
def _check_gids(gids, gid_ranges, valid_cells, arg_name):
```

Just following MNE convention ...

File: hnn_core/check.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def _gid_to_type(gid, gid_ranges):
+    """Reverse lookup of gid to type."""
+    for gidtype, gids in gid_ranges.items():
+        if gid in gids:
+            return gidtype
+
+
+def _check_gid_range(gid, gid_ranges, arg_name):
```
Comment (jasmainak) [inline hnn_core/check.py:43]: I don't see this function used anywhere. Also, I think the `AssertionError` should happen inside `_gid_to_type` so you don't need this anyway

File: examples/howto/plot_connectivity.py [hunk 10]
```diff
@@ -115,16 +120,16 @@ def get_network(probability=1.0):
...
 ###############################################################################
 # We can plot the sparse connectivity pattern between cell populations.
-conn_idx = len(net_sparse.connectivity) - 1
-plot_connectivity_matrix(net_sparse, conn_idx)
-
-conn_idx = len(net_sparse.connectivity) - 2
+src_gids, target_gids = 'L2_basket', 'L2_basket'
+loc, receptor = 'soma', 'gabaa'
+conn_indices = net_erp.pick_connection(src_gids, target_gids, loc, receptor)
+conn_idx = conn_indices[0]
```
Comment (jasmainak) [inline examples/howto/plot_connectivity.py:126]: much better! now only thing is that you shouldn't have to do this:

```suggestion
```

so `plot_connectivity_matrix` accepts a list. Can be a separate PR though. Are there other functions which benefit from `pick_connection` ?

File: hnn_core/network.py [hunk 11]
```diff
@@ -1015,6 +1003,116 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
 
         self.connectivity.append(deepcopy(conn))
 
+    def pick_connection(self, src_gids=None, target_gids=None,
+                        loc=None, receptor=None):
```
Comment (jasmainak) [inline hnn_core/network.py:1007]: I think this would be better off as a function. So that you don't get the impression that calling this method would *modify* `net.connectivity` and include only the connections that satisfy the filter. Pondering over this led me to this stackoverflow post:

https://stackoverflow.com/questions/8108688/in-python-when-should-i-use-a-function-instead-of-a-method

Let me know what you think ...

File: examples/howto/plot_connectivity.py [hunk 12]
```diff
@@ -115,16 +120,16 @@ def get_network(probability=1.0):
 
 ###############################################################################
 # We can plot the sparse connectivity pattern between cell populations.
-conn_idx = len(net_sparse.connectivity) - 1
-plot_connectivity_matrix(net_sparse, conn_idx)
-
-conn_idx = len(net_sparse.connectivity) - 2
+src_gids, target_gids = 'L2_basket', 'L2_basket'
+loc, receptor = 'soma', 'gabaa'
+conn_indices = net_erp.pick_connection(src_gids, target_gids, loc, receptor)
```
Comment (jasmainak) [inline examples/howto/plot_connectivity.py:125]:
```suggestion
conn_indices = net_erp.pick_connection(src_gids='L2_basket', target_gids='L2_basket, loc='soma', receptor='gabaa')
```
I think this is more pythonic as it uses keyword arguments which are less error-prone

**Comment by jasmainak** [comment]
Just some minor comments. Looks fine otherwise!

File: hnn_core/network.py (hunk 11 above)
Comment (ntolley) [reply hnn_core/network.py:1007]: I'm fine with making this a function! I think my main hesitation against this was having to pass the whole network object as an argument, as something seems fishy about passing such a large object to a function with so many irrelevant attributes.

In any case I agree that the idea of a "search function" is better off being separate from the Network, despite being completely tied to the Network object.

File: examples/howto/plot_connectivity.py (hunk 10 above)
Comment (ntolley) [reply examples/howto/plot_connectivity.py:126]: I agree that'd be a bit nicer, but what do you think the alternate behavior should be with a list? Returning multiple plots or subplots?

File: hnn_core/check.py (hunk 9 above)
Comment (ntolley) [reply hnn_core/check.py:43]: I'll just get rid of it. I was thinking of consolidating this error message in the `Network` code, but it's a bit superfluous. Also in some instances we don't want `_gid_to_type` to raise an error.

File: examples/howto/plot_connectivity.py (hunk 10 above)
Comment (jasmainak) [reply examples/howto/plot_connectivity.py:126]: umm ... either is fine. Maybe subplots is better so user is not overwhelmed by plot windows. Each row is one connectivity matrix if multiple subplots. And you can even implement a [scrollbar](https://matplotlib.org/stable/gallery/widgets/slider_demo.html#sphx-glr-gallery-widgets-slider-demo-py) if there are too many subplots

**Review by jasmainak — APPROVED** [review]
PR is good by me

@rythorpe @cjayb do you guys want to take a look?

**Comment by rythorpe** [comment]
Sorry for the delay, I'll try to look look through this tomorrow!

**Comment by jasmainak** [comment]
I went ahead and merged @rythorpe so we can keep the ball rolling. I hope it's okay. Thanks a lot @ntolley for the PR !

**Comment by rythorpe** [comment]
Sounds good, sorry @ntolley for taking so long! This will help a lot with unifying drive and local network connectivity.
